protocol/nats_jetstream/v3/receiver.go

```diff
@@ -0,0 +1,229 @@
+func NewReceiver(ctx context.Context, opts ...ProtocolOption) (*receiver, error) {
+	c := &receiver{
+		incoming:      make(chan msgErr),
+		internalClose: make(chan struct{}, 1),
+	}
+
+	if err := c.applyOptions(opts...); err != nil {
+		return nil, err
+	}
+
+	if err := c.validateOptions(); err != nil {
+		return nil, err
+	}
+
+	var errConnection error
+	defer func() {
+		// close connection if an error occurs and we created the nats connection
+		if c.connOwned && c.conn != nil && errConnection != nil {
+			c.conn.Close()
+		}
+	}()
+
+	// create a connection based on the URL if the connection was not given as an option
+	if c.conn == nil {
+		c.conn, errConnection = nats.Connect(c.url, c.natsOpts...)
+		if errConnection != nil {
+			return nil, errConnection
+		}
+	}
+
+	if c.jetStream, errConnection = jetstream.New(c.conn, c.jetSteamOpts...); errConnection != nil {
+		return nil, errConnection
+	}
+
+	return c, nil
+}
+
+// MsgHandler implements nats.MsgHandler and publishes messages onto our internal incoming channel to be delivered
+// via r.Receive(ctx)
+func (r *receiver) MsgHandler(msg jetstream.Msg) {
+	r.incoming <- msgErr{msg: NewMessage(msg)}
+}
+
+// Receive implements Receiver.Receive.
+func (r *receiver) Receive(ctx context.Context) (binding.Message, error) {
+	select {
+	case msgErr, ok := <-r.incoming:
+		if !ok {
+			return nil, io.EOF
+		}
+		return msgErr.msg, nil
+	case <-ctx.Done():
+		return nil, io.EOF
+	}
+}
+
+// OpenInbound implements Opener.OpenInbound
+func (r *receiver) OpenInbound(ctx context.Context) error {
+	r.subMtx.Lock()
+	defer r.subMtx.Unlock()
+
+	var consumeContext jetstream.ConsumeContext
+	var err error
+	if err = r.createJetstreamConsumer(ctx); err != nil {
+		return err
+	}
+	if consumeContext, err = r.jetstreamConsumer.Consume(r.MsgHandler, r.pullConsumeOpts...); err != nil {
+		return err
+	}
+
+	// Wait until external or internal context done
+	select {
+	case <-ctx.Done():
+	case <-r.internalClose:
+	}
+
+	// Finish to consume messages in the queue and close the subscription
+	if consumeContext != nil {
+		consumeContext.Drain()
+	}
+	return nil
+}
+
+// Close implements Closer.Close
+func (r *receiver) Close(ctx context.Context) error {
+	// Before closing, let's be sure OpenInbound completes
+	// We send a signal to close and then we lock on subMtx in order
+	// to wait OpenInbound to finish draining the queue
+	r.internalClose <- struct{}{}
+	r.subMtx.Lock()
+	defer r.subMtx.Unlock()
+
+	if r.connOwned && r.conn != nil {
+		r.conn.Close()
+	}
+
+	close(r.internalClose)
+
+	return nil
+}
+
+// applyOptions sets fields on the receiver
+func (r *receiver) applyOptions(opts ...ProtocolOption) error {
+	p := &Protocol{receiver: r}
+	for _, fn := range opts {
+		if err := fn(p); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// validateOptions runs after all options have been applied and makes sure needed options were set correctly.
+func (r *receiver) validateOptions() error {
```

Comment:
Please remove this redundant validation logic from the sender/receiver and put it into the protocol. Unless you see a need for additional validation logic in the receiver/sender and not in the higher-level protocol which can do the same, avoid duplication, and return earlier?